chunk/chunk.go

```diff
@@ -21,3 +24,50 @@ type ByID []Chunk
 func (cs ByID) Len() int           { return len(cs) }
 func (cs ByID) Swap(i, j int)      { cs[i], cs[j] = cs[j], cs[i] }
 func (cs ByID) Less(i, j int) bool { return cs[i].ID < cs[j].ID }
+
+// ChunksToMatrix converts a slice of chunks into a model.Matrix.
+func ChunksToMatrix(chunks []Chunk) (model.Matrix, error) {
+	// Group chunks by series, sort and dedupe samples.
+	sampleStreams := map[model.Fingerprint]*model.SampleStream{}
+
+	for _, c := range chunks {
+		fp := c.Metric.Fingerprint()
+		ss, ok := sampleStreams[fp]
+		if !ok {
+			ss = &model.SampleStream{
+				Metric: c.Metric,
+			}
+			sampleStreams[fp] = ss
+		}
+
+		samples, err := decodeChunk(c.Data)
+		if err != nil {
+			return nil, err
+		}
+
+		ss.Values = util.MergeSamples(ss.Values, samples)
```

Comment:
Merging lots of small lists into a large list will be n^2 in the worse case; ideally we'd to a merge-sort style tree merge. Not need to do it now, but file a ticket?